src/main/java/hu/oe/nik/szfmv/automatedcar/xml/converter/ObjectConverter.java

```diff
@@ -0,0 +1,57 @@
+package hu.oe.nik.szfmv.automatedcar.xml.converter;
+
+import hu.oe.nik.szfmv.automatedcar.model.*;
+
+import javax.xml.bind.annotation.adapters.XmlAdapter;
+
+import javax.xml.bind.Binder;
+import javax.xml.bind.JAXBContext;
+import javax.xml.bind.JAXBElement;
+
+import hu.oe.nik.szfmv.automatedcar.xml.XmlParser;
+import org.apache.commons.lang3.StringUtils;
+import org.w3c.dom.Node;
+
+/**
+ * Nem sikerült mindent annotációval megoldani, ezért jött létre ez a segédosztály,
+ * ami megmondja a JAXB-nek, hogy mit példányosítson.
+ */
+public class ObjectConverter extends XmlAdapter<Object, WorldObject> {
+
+    @Override
+    public WorldObject unmarshal(Object element) throws Exception {
+        Node node = (Node)element;
+        String classType = node.getAttributes().getNamedItem("type").getNodeValue();
+        Class<?> clazz = getClassByString(classType);
+
+        JAXBContext jaxbContext = XmlParser.getJAXBContextFromCache(clazz);
+        Binder<Node> binder = jaxbContext.createBinder();
+        JAXBElement<?> jaxBElement = binder.unmarshal(node, clazz);
+        return (WorldObject) jaxBElement.getValue();
+    }
+
+    private Class<?> getClassByString(String classType) {
+        if (StringUtils.startsWith(classType,"road_2")){
+            return Road.class;
+        } else if (StringUtils.startsWith(classType,"tree")) {
+            return Tree.class;
+        } else if (StringUtils.startsWith(classType,"parking_space")) {
+            return ParkingSpace.class;
+        } else if (StringUtils.startsWith(classType,"crosswalk")) {
+            return Crosswalk.class;
+        } else if (StringUtils.startsWith(classType,"roadsign")) {
+            return Sign.class;
+        } else if (StringUtils.startsWith(classType, "car")){
+            return AutomatedCar.class;
+        }  else if (StringUtils.startsWith(classType, "parking_bollard")) {
+            return ParkingBollard.class;
+        } else { // a fail-safe működés miatt került be, lehetne exceptiont is dobni
+            return WorldObject.class;
```

Comment:
TODO-t pakolj rá pls

Reply:
Közben átgondoltam, nem szabad exception-t dobni, mert akkor az egész beolvasás elszáll. Úgy próbáltam megírni az egész folyamatot, hogy hibás XML-re ne dobjon hibát, hanem kezelje (meg persze logolja) és amit tud, olvasson be. Egy error logot kellene berakni, az tuti. Átgondolom még mi lenne elegáns megoldás itt, de annyira nem érzem égető feladatnak most, lehet tényleg pakolok rá egy TODO-t.

Reply:
Szerintem mehet exception is akár. Valójában milyen hibás XML-re gondoltok? Egy invalid XML-t (hiányzó lezáró tag pl.) nem kell megpróbálni feloldani, elszállhat az egész. Az általunk biztosított XML valid. Nem is fogunk megpróbálni invalid XML-t betölteni. Ez a tárgy nem ilyesmiről szól.

Az XML-t egy Bosch-os belső tool generálja, bár az az XML gusztustalan, így még áteresztjük egy szkripten ami kidobja belőle a szemetet. De annak a kimenetét is validnak tekintjük mind szintaktikailag mind szemantikailag. (Eltekintve az amúgy sem használt road painting node-októl :D)

Igazából hirtelen nem látok olyan XML parse-olási hibaforrást ami az XML-ből eredhetne.

Reply:
Igazából itt arra gondolunk ha olyan objektum kerül az xmlbe ami nincs benne a nagy if-else szörnyetegben, a logolás elegáns szerintem, mert attól még minden más bekerül és a logból látjuk majd, hogy kell egy extra osztály

Reply:
A hibás XML-t úgy értettem, hogy ismeretlen type kerül be, kicsit félreérthető volt. :D Ha a fájlban pl. hiányzó lezáró tag lenne, akkor szerintem már a fájl beolvasásnál elszállna.

Persze én is azt feltételeztem, hogy nem fogjuk feszegetni az XML beolvasás határai, bár már lehetne, mert szerintem sok mindent kezel. :D Az ehhez tartozó teszt is ellenőrzi az else ágat, mert van egy "unknown_type" típusú elem a teszt XML-ben. Az eredeti ok, amiért beraktam, hogy először a 300 soros XML-el teszteltem, aztán beolvastam az 5000 sorosat is, amiben szerepel parking_bollard, ami a kicsiben nem található, így kaptam egy exceptiont. Innen jött az ötlet, hogy ilyet nem szeretnék megengedni a későbbiekben. 

Ugyanakkor ha a másik irányból közelítünk, akkor lesz egy olyan elem a WorldObjectek között, ami nem egy valid objektum, mégis létrehozott egy WO-t, amit berakott a világba.